### ISSUE TITLE
Make the compiler runtime independent of the standard library

### ISSUE THREAD

**Issue by odersky** [lampepfl/dotty#5698]

**Comment by hepin1989** [comment]
The classes of the kernel library will be visible or not? AFAIK, SBT itself can let us filter out the scala-library with `autoScalaLibrary := false`. For me I think without a must in `scala-library` dependency will make scala very useful on many things. eg, the latest very good usage is how to write a library in scala which will not depend on `scala-library` itself in https://github.com/lightbend/config/pull/600#issuecomment-443565128

**Comment by huntc** [comment]
My first reaction is to the memory footprint of an application. Most of my Scala apps depend on some Java library. Therefore, the idea of an independent runtime makes me wonder if the memory footprint will increase given that Java’s runtime would be linked.

Related: I think that the Scala and JVM ecosystems at large need to be sensitive to their memory footprint, particularly due the rise of Go and Rust.

**Comment by odersky** [comment]
@hepin1989 The classes of kernel library would be visible. TupleN, FunctionN, StringContext are all programmer accessible.

@huntc The Java libraries would still be included (on JVM at least, would be different for Scala.js and Scala.native). The only thing that would change is that you could write Scala programs that do not rely on the Scala library, if you are careful.

**Comment by propensive** [comment]
What else? `Symbol`s and boxed types spring to mind.

Extension methods defined in the standard library could even provide most of the (currently familiar) methods of these "kernel" classes so they wouldn't need to be "fully featured" implementations.

**Comment by ghik** [comment]
Does dotty still use `Seq` to express varargs? Does that mean that a lot of collection library would have to be included?

**Comment by lihaoyi** [comment]
I think this would make a big difference for Scala.js; currently it's not feasible to use Scala.js for tiny scripts because it has to pull in and optimize the entire std lib, meaning a de-facto minimum size of 100kb or so. Having std-lib-free Scala code would both make the output executables tiny for the cases where that matters (bookmarklets, microcontrollers, etc) and also make the output code much easier to skim (since you don't need to mentally filter out the leftover bits of the entire collections library).

This would also make it more feasible to use Scala for tiny bootstrap scripts; current Mill's launcher is written in Java because the added classloading needed to use scala.Predef (even just `println`) easily adds a 200-400ms of initialization overhead. Would be nice to be able to write it in Scala even if that means avoiding all those collections

**Comment by smarter** [comment]
@lihaoyi Scala.JS already does all the DCE I can do, so in theory you should already be able to achieve this just by carefully avoiding using too much stuff from the standard library (which would probably require developing your own mini standard library), I don't think the split proposed here would make a difference (/cc @sjrd who can correct me if I'm wrong here).

**Comment by smarter** [comment]
> What else? 

See this fantastic SO answer by Seb: https://stackoverflow.com/questions/40076047/what-types-are-special-to-the-scala-compiler/40083398#40083398

**Comment by sjrd** [comment]
> Scala.JS already does all the DCE it can do, so in theory you should already be able to achieve this just by carefully avoiding using too much stuff from the standard library

Yes, if you stick to JS types, never use Scala functions or collections, then Scala.js will dce the standard library away. Once fullOpt'ed, the smallest Scala.js program (only a `main` with `js.Dynamic.global.console.log("hello")`, *not* `println`) weighs 8 KB in Scala.js 1.x master. Most of the code in there is about the implementation of `Long`s and some metadata for boxed classes and `java.lang.Class`. Not a single Scala class in sight (except `scala.runtime.BoxedUnit`, but [that could change](https://github.com/scala-js/scala-js/pull/3535)).

**Comment by dwijnand** [comment]
I agree it would be a wonderful achievement (one I've even pushed for at one point, using the nomenclature "core" rather than "kernel").

But I think, even if the compiler's generated code doesn't depend on it, the standard library defines the language just as much as the compiler does.  I mention that because I'm not sure I understand the benefits, asides from the footprint one.

The one benefit that I would love to see is a kernel library that is backwards binary-compatible across major releases, so scala-kernel.jar, not scala-kernel_3.0.jar.  That would allow creating libraries in Scala that are:
1. cross-platform (JVM/Scala.js/ScalaNative), and
2. Java-friendly, by not pulling in the problematic scala library

**Comment by ghik** [comment]
> Well, that's exactly what this proposal is isn't it? Having everyone agree on a mini standard library ("kernel") so people can know what to use to avoid using stuff from the standard library.
I think what @smarter meant was that when you don't use *standard* standard library then you'll end up reimplementing parts of it (e.g. collections) in a new, mini standard library.

**Comment by ghik** [comment]
> Make vararg parameters take an immutable array instead of a Seq. This has the potential of
simplifying Java interop at the price of complicating interop with 2.13.

Maybe another option is to introduce an abstraction for repeated parameters, i.e. a `trait Repeated[+A]`? It would have a minimal, array-like interface so that it would not require collections library. At the same time `Seq` would inherit from `Repeated` so that collections could be passed into varargs methods without copying them into immutable arrays.
